kover-gradle-plugin/src/main/kotlin/kotlinx/kover/gradle/aggregation/project/KoverProjectGradlePlugin.kt

```diff
@@ -0,0 +1,181 @@
+import kotlinx.kover.gradle.aggregation.commons.names.PluginId.KOTLIN_MULTIPLATFORM_PLUGIN_ID
+import kotlinx.kover.gradle.aggregation.commons.names.SettingsNames
+import kotlinx.kover.gradle.aggregation.commons.utils.bean
+import kotlinx.kover.gradle.aggregation.commons.utils.hasSuper
+import kotlinx.kover.gradle.aggregation.project.instrumentation.InstrumentationFilter
+import kotlinx.kover.gradle.aggregation.project.instrumentation.JvmOnFlyInstrumenter
+import kotlinx.kover.gradle.aggregation.project.tasks.ArtifactGenerationTask
+import kotlinx.kover.gradle.aggregation.project.tasks.KoverAgentSearchTask
+import kotlinx.kover.gradle.plugin.commons.KoverCriticalException
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.api.Task
+import org.gradle.api.file.ConfigurableFileCollection
+import org.gradle.api.tasks.compile.JavaCompile
+import org.gradle.api.tasks.testing.Test
+import org.gradle.kotlin.dsl.create
+import org.gradle.kotlin.dsl.register
+import org.gradle.kotlin.dsl.withType
+import java.io.File
+
+internal class KoverProjectGradlePlugin : Plugin<Project> {
+
+    override fun apply(target: Project) {
+        if (target.path == Project.PATH_SEPARATOR) {
+            target.configureAgentSearch()
+        }
+
+        target.configureInstrumentation()
+        target.configureArtifactGeneration()
+    }
+
+    private fun Project.configureInstrumentation() {
+        val koverJarDependency = configurations.getByName(SettingsNames.DEPENDENCY_AGENT)
+        val jarConfig = configurations.create("agentJarSource") {
+            asConsumer()
+            attributes {
+                attribute(KoverContentAttr.ATTRIBUTE, KoverContentAttr.AGENT_JAR)
+            }
+            extendsFrom(koverJarDependency)
+        }
+        JvmOnFlyInstrumenter.instrument(tasks.withType<Test>(), jarConfig, InstrumentationFilter(setOf(), setOf()))
+    }
+
+    private fun Project.configureArtifactGeneration() {
+        val taskGraph = gradle.taskGraph
+
+        val artifactFile = layout.buildDirectory.file("kover/kover.artifact")
+
+        // we create task immediately because of mustRunAfter
+        val generateArtifactTask = tasks.create<ArtifactGenerationTask>("koverGenerateArtifact")
+        generateArtifactTask.outputFile.set(artifactFile)
+
+        // add tests
+        val testTasks = tasks.withType<Test>().matching { task ->
+            taskGraph.hasTask(task.path)
+        }
+
+        val binReportFiles = project.layout.buildDirectory.dir(binReportsRootPath())
+            .map { dir -> testTasks.map { dir.file(binReportName(it.name)) } }
+
+        val exts = extensions
+        val pluginManager = pluginManager
+        val projectPath = path
+
+        val compilations = project.layout.buildDirectory.map {
+            val compilations = when {
+                pluginManager.hasPlugin(KOTLIN_JVM_PLUGIN_ID) -> {
+                    val kotlin = exts.findByName("kotlin")?.bean()
```

Comment:
Nice trick with the DynamicBean!

Reply:
I don't like dynamic objects (
But this is the easiest way to get around the problems with different class loaders.